src/ring_theory/non_unital_subsemiring/basic.lean

```diff
@@ -0,0 +1,820 @@
+/-
+Copyright (c) 2022 Jireh Loreaux All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+
+import algebra.module.basic
+import algebra.ring.equiv
+import algebra.ring.prod
+import data.set.finite
+import group_theory.submonoid.centralizer
+import group_theory.submonoid.membership
+import group_theory.subsemigroup.membership
+
+/-!
+# Bundled non-unital subsemirings
+
+We define bundled non-unital subsemirings and some standard constructions:
+`complete_lattice` structure, `subtype` and `inclusion` ring homomorphisms, non-unital subsemiring
+`map`, `comap` and range (`srange`) of a `non_unital_ring_hom` etc.
+-/
+
+open_locale big_operators
+
+universes u v w
+
+variables {R : Type u} {S : Type v} {T : Type w} [non_unital_non_assoc_semiring R]
+  (M : subsemigroup R)
+
+section non_unital_subsemiring_class
+
+/-- `non_unital_subsemiring_class S R` states that `S` is a type of subsets `s ⊆ R` that
+are both an additive submonoid and also a multiplicative subsemigroup. -/
+class non_unital_subsemiring_class (S : Type*) (R : out_param $ Type u)
+  [non_unital_non_assoc_semiring R] [set_like S R] extends add_submonoid_class S R :=
+(mul_mem : ∀ {s : S} {a b : R}, a ∈ s → b ∈ s → a * b ∈ s)
+
+@[priority 100] -- See note [lower instance priority]
+instance non_unital_subsemiring_class.mul_mem_class (S : Type*) (R : out_param $ Type u)
+  [non_unital_non_assoc_semiring R] [set_like S R] [h : non_unital_subsemiring_class S R] :
+  mul_mem_class S R :=
+{ .. h }
+
+variables [set_like S R] [hSR : non_unital_subsemiring_class S R] (s : S)
+include hSR
+
+open add_submonoid_class
+
+namespace non_unital_subsemiring_class
+
+/-- A non-unital subsemiring of a `non_unital_non_assoc_semiring` inherits a
+`non_unital_non_assoc_semiring` structure -/
+@[priority 75] /- Prefer subclasses of `non_unital_non_assoc_semiring` over subclasses of
+`non_unital_subsemiring_class`. -/
+instance to_non_unital_non_assoc_semiring : non_unital_non_assoc_semiring s :=
+subtype.coe_injective.non_unital_non_assoc_semiring coe rfl (by simp) (λ _ _, rfl) (λ _ _, rfl)
+
+instance no_zero_divisors [no_zero_divisors R] : no_zero_divisors s :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero := λ x y h,
+  or.cases_on (eq_zero_or_eq_zero_of_mul_eq_zero $ subtype.ext_iff.mp h)
+    (λ h, or.inl $ subtype.eq h) (λ h, or.inr $ subtype.eq h) }
```

Comment:
I think we have `function.injective.no_zero_divisors` for this (invoked with `subtype.coe_injective`)